### PR TITLE
Add default config properties to chains configMap

### PR DESCRIPTION
**Pull request by jkhelil**
# Changes
- Add chains option to allow users to opt-in to generate a chains signing secret and generate and ecdsa key pair as signing secret
- Addition of Sane Default Configuration Properties:
This update introduces sensible default configuration properties to the Chains controller, enhancing its initial setup and performance.
- Storage Backend Update:
Previously, the Chains controller used Tekton storage, where payloads and signatures were stored in the annotations of Tekton objects (such as TaskRuns and PipelineRuns). This approach significantly increased the size of these objects. To address this issue, this pull request changes the default storage backend to OCI, thereby reducing the object size and improving overall efficiency.
By default chains controller uses storage type tekton for its differents added annotation

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Run `make test lint` before submitting a PR
- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/operator/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

### CODE REVIEW

**Comment by jkhelil** [comment]
@PuneetPunamiya got my old PR closed, changed the branch name
Anyway, I took in account your  useful comments, and did fixes here
Now whenever we change the geenratedSigningSecret input, it deletes the installerset
However i am still seeing an issue, with tektonconfig defaults, those are lending correctly on the chains configmap, but once i update one of them in tektonconfig, they dont get updated on chains configmap. Wondering if this is correct place where we should put them, @piyush-garg ?

File: pkg/apis/operator/v1alpha1/tektonconfig_defaults.go [hunk 1]
```diff
@@ -94,4 +94,24 @@ func (tc *TektonConfig) SetDefaults(ctx context.Context) {
...
+	}
+	if tc.Spec.Chain.ArtifactsPipelineRunStorage == nil {
+		tc.Spec.Chain.ArtifactsPipelineRunStorage = ptr.String("oci")
+	}
+	if tc.Spec.Chain.ArtifactsOCIFormat == "" {
+		tc.Spec.Chain.ArtifactsOCIFormat = "simplesigning"
+	}
+	if tc.Spec.Chain.ArtifactsOCIStorage == nil {
+		tc.Spec.Chain.ArtifactsOCIStorage = ptr.String("oci")
+	}
```
Comment (jkandasa) [inline pkg/apis/operator/v1alpha1/tektonconfig_defaults.go:116]: @jkhelil these defaulting will not impact existing chains CR, 
However, if user updates a field on a TektonConfig CR, not related to chains also changes the chains behavior.
Will that impact the existing chains instance?

File: pkg/reconciler/common/transformers.go [hunk 2]
```diff
@@ -940,3 +940,49 @@ func ReplaceNamespace(newNamespace string) mf.Transformer {
...
+
+		// Convert unstructured to Secret
+		secret := &corev1.Secret{}
+		err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, secret)
+		if err != nil {
+			return err
+		}
+
+		// Update the Secret's data only if it is nil or empty
+		if secret.Data == nil || len(secret.Data) == 0 {
```
Comment (jkandasa) [inline pkg/reconciler/common/transformers.go:965]: @jkhelil I believe `len` should return `0` on `nil` value too. If so, we can remove the `nil` check

File: pkg/reconciler/kubernetes/tektonchain/tektonchain.go [hunk 3]
```diff
@@ -46,6 +52,9 @@ const (
 	chainsContainerName = "tekton-chains-controller"
 	// Deployment Name
 	chainsDeploymentName = "tekton-chains-controller"
+
+	// secret installer set additional Annotation
+	SecretTISSigningAnnotation = "operator.tekton.dev/generated-signing-secret"
```
Comment (jkandasa) [inline pkg/reconciler/kubernetes/tektonchain/tektonchain.go:57]: if we do not use `SecretTISSigningAnnotation` outside package, we can make it `private`

File: pkg/reconciler/kubernetes/tektonchain/tektonchain.go [hunk 4]
```diff
@@ -469,6 +478,27 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, tc *v1alpha1.TektonChain
 		}
 		return nil
 	}
+	// if generatedSigningSecret has been changed for chainsCR, then update the Tekton Chain Secret InstallerSet
+	secretInstallerSetSigningKey, _ := strconv.ParseBool(installedSecretTIS.Annotations[SecretTISSigningAnnotation])
```
Comment (jkandasa) [inline pkg/reconciler/kubernetes/tektonchain/tektonchain.go:482]: Does `SecretTISSigningAnnotation` present always in the annotations? on the existing running instance too?

**Review by jkandasa — COMMENTED** [review]
@jkhelil Please confirm the upgrade happens smoothly from the previous version of chains

File: pkg/reconciler/kubernetes/tektonchain/tektonchain.go (hunk 4 above)
Comment (jkhelil) [reply pkg/reconciler/kubernetes/tektonchain/tektonchain.go:482]: No it is added if the secret is generated

File: pkg/apis/operator/v1alpha1/tektonconfig_defaults.go (hunk 1 above)
Comment (jkhelil) [reply pkg/apis/operator/v1alpha1/tektonconfig_defaults.go:116]: yes i guess, it is going to set defaults, but thats what we expect, we do want to set these defaults

Comment (piyush-garg) [reply pkg/apis/operator/v1alpha1/tektonconfig_defaults.go:116]: this defaulting should be done in chains CR, and from here we should be just calling that

File: pkg/reconciler/kubernetes/tektonchain/tektonchain.go (hunk 4 above)
Comment (jkandasa) [reply pkg/reconciler/kubernetes/tektonchain/tektonchain.go]: If the annotation `secretTISSigningAnnotation` not present, return empty string ``, that leads bool parse error,
```
false strconv.ParseBool: parsing "": invalid syntax
```

File: pkg/apis/operator/v1alpha1/tektonchain_defaults_test.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+	"context"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/tektoncd/pipeline/test/diff"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/pkg/ptr"
+)
+
+func test_SetDefaults_Chain(t *testing.T) {
```
Comment (jkandasa) [inline pkg/apis/operator/v1alpha1/tektonchain_defaults_test.go:29]: @jkhelil can you camelCase the function name? ==> `testSetDefaultsChain`

error:
```
pkg/apis/operator/v1alpha1/tektonchain_defaults_test.go:29:6: func `test_SetDefaults_Chain` is unused (unused)
func test_SetDefaults_Chain(t *testing.T) {
```

File: pkg/apis/operator/v1alpha1/tektonchain_defaults_test.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
+/*
+Copyright 2022 The Tekton Authors
```
Comment (jkandasa) [inline pkg/apis/operator/v1alpha1/tektonchain_defaults_test.go:2]: year should be 2024

File: docs/TektonChain.md [hunk 7]
```diff
@@ -113,6 +121,14 @@ spec:
   transparency.enabled: #value (boolean - true/false)
   transparency.url: #value
 ```
+- `disabled` : if the value set as `true`, chains feature will be disabled (default: `false`)
+- `generateSigningSecret`: If set to true, the operator will generate an ECDSA key pair (x509.pem private key and x509-pub.pem public key) and store it in the "signing-secrets" secret in the "tekton-pipelines" namespace. the secret is used by chains controller to sign tekton artifacts(tasjruns, pipelineruns). it is important to mention that:
```
Comment (jkandasa) [inline docs/TektonChain.md:125]: typo: `tasjruns`

**Review by jkandasa — APPROVED** [review]

**Pull request merged**